### PR TITLE
Fail if output memory not allocated

### PR DESCRIPTION
**Pull request by lamblin**
instead of segfaulting like a barbarian.
Also add a couple of exceptions before failures.

### CODE REVIEW

File: theano/gpuarray/subtensor.py [hunk 1]
```diff
@@ -107,7 +107,10 @@ def c_code(self, node, name, inputs, outputs, sub):
             return """
         Py_XDECREF(%(out)s);
         %(out)s = pygpu_copy(%(inp)s, GA_ANY_ORDER);
-        if (!%(out)s) { %(fail)s }
+        if (!%(out)s) {
+            PyErr_SetString(PyExc_MemoryError, "Could not allocate copy of input.");
```
Comment (abergeron) [inline theano/gpuarray/subtensor.py]: There should already be a proper exception with a suitable message here.

File: theano/gpuarray/subtensor.py [hunk 2]
```diff
@@ -733,6 +736,7 @@ def c_code(self, node, name, inputs, outputs, sub):
         if (!%(inplace)s) {
           %(out)s = theano_try_copy(%(out)s, %(x)s);
           if (%(out)s == NULL)
+            PyErr_SetString(PyExc_MemoryError, "Could not allocate copy of input.");
```
Comment (abergeron) [inline theano/gpuarray/subtensor.py]: There is already an exception.

File: theano/gpuarray/subtensor.py [hunk 3]
```diff
@@ -874,6 +878,7 @@ def c_code(self, node, name, inputs, outputs, sub):
   %(out)s = theano_try_copy(%(out)s, %(x)s);
 }
 if (!%(out)s) {
+  PyErr_SetString(PyExc_MemoryError, "Could not allocate copy of input.");
```
Comment (abergeron) [inline theano/gpuarray/subtensor.py]: There is already an exception.

File: theano/tensor/subtensor.py [hunk 4]
```diff
@@ -1408,6 +1408,10 @@ def c_code(self, node, name, inputs, outputs, sub):
         {
             Py_XDECREF(%(z)s);
             %(z)s = %(copy_of_x)s;
+            if (!%(z)s) {
+                PyErr_SetString(PyExc_MemoryError, "Could not allocate copy of input.");
```
Comment (abergeron) [inline theano/tensor/subtensor.py]: I am reasonably certain that the CPU code already sets an exception and 100% certain that the GPU code does.

File: theano/tensor/subtensor.py [hunk 5]
```diff
@@ -1972,6 +1978,10 @@ def c_code(self, node, name, input_names, output_names, sub):
         {
             Py_XDECREF(%(out)s);
             %(out)s = %(copy_of_x)s;
+            if (!%(out)s) {
+                PyErr_SetString(PyExc_MemoryError, "Could not allocate copy of input.");
```
Comment (abergeron) [inline theano/tensor/subtensor.py]: I am reasonably certain that the CPU code already sets an exception and 100% certain that the GPU code does.

**Comment by abergeron** [comment]
Weirdly, with all the messages you added, jenkins complains that there is an exception missing.

File: theano/gpuarray/subtensor.py (hunk 1 above)
Comment (lamblin) [inline theano/gpuarray/subtensor.py]: OK.

File: theano/gpuarray/subtensor.py (hunk 2 above)
Comment (lamblin) [inline theano/gpuarray/subtensor.py]: Oh, I'm missing curly braces there, which would explain why things get weird.
I'll fix that later.

**Comment by lamblin** [comment]
Updated and squashed.

**Comment by lamblin** [comment]
While I'm playing in that code, I added a fix for 6.0 arch.

**Pull request merged**